src/backend/lib/types.d.ts

```diff
@@ -0,0 +1,111 @@
+// @ts-check
+
+import { Static } from "@sinclair/typebox";
+import { InstallerExtSchema } from "./schema";
+import {
+  FastifyReply,
+  FastifyRequest,
+  RawServerDefault,
+  RawReplyDefaultExpression,
+  RawRequestDefaultExpression,
+} from "fastify";
+import { PartialDeep } from "type-fest";
+import UpgradeManager from "./upgrade-manager";
+
+// External installer type, in server responses
+export type InstallerExt = Static<typeof InstallerExtSchema>;
+// Internal installer type (for installers stored locally), with filepath instead of url
+export type InstallerInt = Omit<InstallerExt, "url"> & {
+  filepath: string;
+};
+export type Reply<RouteGeneric> = FastifyReply<
+  RawServerDefault,
+  RawRequestDefaultExpression,
+  RawReplyDefaultExpression,
+  RouteGeneric
+>;
+export type Request<RouteGeneric> = FastifyRequest<
+  RouteGeneric,
+  RawServerDefault,
+  RawRequestDefaultExpression
+>;
+// An upload or a download in progress
+export interface TransferProgress {
+  /** id (hash) of the file being transferrer */
+  id: string;
+  /** bytes transferred so far */
+  sofar: number;
+  /** total number of bytes to transfer */
+  total: number;
+}
+// Info about the current device
+export interface DeviceInfo {
+  supportedAbis: InstallerExt["arch"];
```

Comment:
What does this syntax do?

Reply:
This selects the type of the property `arch` from the object/interface `InstallerExt`. Typescript does not support "dot" notation to access properties like you would on a Javascript object (e.g. `myObject.prop`, so you have to use `myObject["prop"]` when referring to props.

TL;DR this says that the `supportedAbis` prop has the same type as the `arch` prop of the `InstallerExt` type.